src/imperative/imperative.cc

```diff
@@ -316,181 +308,220 @@ std::vector<NDArray*> Imperative::Backward(
-      states[idx.node_id(n.get())] = info.state;
-      for (uint32_t i = 0; i < info.outputs.size(); ++i) {
-        CHECK(idx.exist(n.get()));
-        size_t nid = idx.node_id(n.get());
-        size_t eid = idx.entry_id(nid, i);
+    nnvm::DFSVisit(forward_outputs, [&](const nnvm::NodePtr& n) {
+      const AGInfo& info = AGInfo::Get(n);
+      states.at(indexed_graph.node_id(n.get())) = info.state;
+      for (size_t i = 0; i < info.outputs.size(); ++i) {
+        CHECK(indexed_graph.exist(n.get()));
+        const size_t nid = indexed_graph.node_id(n.get());
+        const size_t eid = indexed_graph.entry_id(nid, i);
         buff[eid] = info.outputs[i];
-        buff[eid].entry_ = NodeEntry{n, i, 0};
+        buff[eid].autograd_ = NodeEntry{n, static_cast<uint32_t>(i), 0};
         ref_count[eid] = 1;
       }
     });
     for (auto& ograd_entry : ograd_entries) {
-      AGInfo& info = AGInfo::Get(ograd_entry.node);
-      if (!idx.exist(ograd_entry.node.get())) continue;
-      size_t eid = idx.entry_id(ograd_entry);
+      const AGInfo& info = AGInfo::Get(ograd_entry.node);
+      if (!indexed_graph.exist(ograd_entry.node.get())) continue;
+      size_t eid = indexed_graph.entry_id(ograd_entry);
       buff[eid] = info.outputs[0];
-      buff[eid].entry_ = ograd_entry;
+      buff[eid].autograd_ = ograd_entry;
     }
   } else {
     states.reserve(num_forward_nodes);
     for (size_t i = 0; i < num_forward_nodes; ++i) {
-      const AGInfo& info = dmlc::get<AGInfo>(idx[i].source->info);
+      // TODO(larroy): This is a code smell 💩
+      AGInfo& info = const_cast<AGInfo&>(dmlc::get<AGInfo>(indexed_graph[i].source->info));
       states.emplace_back(info.state);
       for (size_t j = 0; j < info.outputs.size(); ++j) {
-        size_t eid = idx.entry_id(i, j);
-        arrays[eid] = const_cast<NDArray*>(&(info.outputs[j]));
-
-        if (retain_graph || info.grad_req != kNullOp) ref_count[eid] = 1;
+        const size_t eid = indexed_graph.entry_id(i, j);
+        arrays[eid] = &(info.outputs[j]);
+        if (retain_graph || info.grad_req != kNullOp)
+          ref_count[eid] = 1;
       }
     }
     for (auto& ograd_entry : ograd_entries) {
-      if (!idx.exist(ograd_entry.node.get())) continue;
+      if (!indexed_graph.exist(ograd_entry.node.get())) continue;
       AGInfo& info = AGInfo::Get(ograd_entry.node);
-      arrays[idx.entry_id(ograd_entry)] = &info.outputs[0];
+      arrays[indexed_graph.entry_id(ograd_entry)] = &info.outputs[0];
     }
   }
   for (size_t i = num_forward_outputs; i < graph.outputs.size(); ++i) {
-    size_t eid = idx.entry_id(graph.outputs[i]);
-    arrays[eid] = x_grads[i - num_forward_outputs];
+    size_t eid = indexed_graph.entry_id(graph.outputs[i]);
+    arrays[eid] = gvars.gradients[i - num_forward_outputs];
     ref_count[eid] = 1;
   }
 
   // Assign context
-  auto vctx = PlaceDevice(idx);
+  auto vctx = PlaceDevice(indexed_graph);
 
   // Infer shape type
   {
```

Comment:
Can we create a function to infershape here, instead of using braces.

Reply:
Since this work itself is a refactoring, would it be better to perform a complete one?